pkg/cloud/baremetal/actuators/machine/actuator.go

```diff
@@ -584,3 +593,192 @@ func (a *Actuator) nodeAddresses(host *bmh.BareMetalHost) ([]corev1.NodeAddress,
+func (a *Actuator) requestPowerOff(ctx context.Context, baremetalhost *bmh.BareMetalHost) error {
+	if baremetalhost.Annotations == nil {
+		baremetalhost.Annotations = make(map[string]string)
+	}
+
+	if _, powerOffRequestExists := baremetalhost.Annotations[requestPowerOffAnnotation]; powerOffRequestExists {
+		return &clustererror.RequeueAfterError{RequeueAfter: time.Second * 5}
+	}
+
+	baremetalhost.Annotations[requestPowerOffAnnotation] = ""
+
+	err := a.client.Update(ctx, baremetalhost)
+	if err != nil {
+		log.Printf("failed to add power off request annotation to %s: %s", baremetalhost.Name, err.Error())
+	}
+
+	return err
+}
+
+//requestPowerOn removes requestPowerOffAnnotation from baremetalhost which signals BMO to power on the machine
+func (a *Actuator) requestPowerOn(ctx context.Context, baremetalhost *bmh.BareMetalHost) error {
+	if baremetalhost.Annotations == nil {
+		baremetalhost.Annotations = make(map[string]string)
+	}
+
+	if _, powerOffRequestExists := baremetalhost.Annotations[requestPowerOffAnnotation]; !powerOffRequestExists {
+		return &clustererror.RequeueAfterError{RequeueAfter: time.Second * 5}
+	}
+
+	delete(baremetalhost.Annotations, requestPowerOffAnnotation)
+
+	err := a.client.Update(ctx, baremetalhost)
+	if err != nil {
+		log.Printf("failed to power-off request annotation from %s: %s", baremetalhost.Name, err.Error())
+	}
+
+	return err
+}
+
+// deleteMachineNode deletes the node that mapped to specified machine
+func (a *Actuator) deleteNode(ctx context.Context, node *corev1.Node) error {
+	if !node.DeletionTimestamp.IsZero() {
+		return &clustererror.RequeueAfterError{RequeueAfter: time.Second*2}
+	}
+
+	err := a.client.Delete(ctx, node)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return &clustererror.RequeueAfterError{}
+		}
+		log.Printf("Failed to delete node %s: %s", node.Name, err.Error())
+		return err
+	}
+	return &clustererror.RequeueAfterError{RequeueAfter: time.Second*2}
+}
+
+// getNodeByMachine returns the node object referenced by machine
+func (a *Actuator) getNodeByMachine(ctx context.Context, machine *machinev1.Machine) (*corev1.Node, error) {
+	if machine.Status.NodeRef == nil {
+		return nil, errors.NewNotFound(corev1.Resource("ObjectReference"), machine.Name)
+	}
+
+	node := &corev1.Node{}
+	key := client.ObjectKey{
+		Name:      machine.Status.NodeRef.Name,
+		Namespace: machine.Status.NodeRef.Namespace,
+	}
+
+	if err := a.client.Get(ctx, key, node); err != nil {
+		return nil, err
+	}
+	return node, nil
+}
+
+/*
+remediateIfNeeded will try to remediate unhealthy machines (annotated by MHC) by power-cycle the host
+The full remediation flow is:
+1) Power off the host
+2) Add poweredOffForRemediation annotation to the unhealthy Machine
+3) Delete the node
+4) Power on the host
+5) Wait for the node the come up
+6) Remove poweredOffForRemediation annotation and the annotation added by MAO to signal the machine is unhealthy
```

Comment:
This can be read two different ways. Suggest something like: "Remove poweredOffForRemediation annotation and the MAO's machine unhealthy annotation."

Reply:
changed